### ISSUE TITLE
Log epoch metrics before the `on_evaluation_end` hook

### ISSUE THREAD

**Issue by dcfidalgo** [PyTorchLightning/pytorch-lightning#7166]
## 🚀 Feature
I would like the epoch metrics to get logged before the `on_evaluation_end` hook.

### Motivation

When doing HPO with the [Ray Tune](https://docs.ray.io/en/master/tune/index.html) library, you usually report the metrics back to Tune `on_validation_end`. If you use a trial scheduler such as the ASHA algorithm, it will terminate the trial if the metric indicates low performance for this specific trial. If i want to visualize the HPO in WandB, for example, i would like to know for which values of the metric the trials got terminated. At the moment this is not possible, since the epoch metrics are not logged before the trial gets terminated.

### Pitch

Right now the order of hooks and logging is this one (taken from the `Trainer.run_evaluation()` method, version 1.3.0rc1):
```python
        # hook
        self.evaluation_loop.on_evaluation_epoch_end()

        # update epoch-level lr_schedulers
        if on_epoch:
            self.optimizer_connector.update_learning_rates(interval='epoch')

        # hook
        self.evaluation_loop.on_evaluation_end()

        # log epoch metrics
        eval_loop_results = self.evaluation_loop.log_epoch_metrics_on_evaluation_end()
```
Simply switching the order of the `log_epoch_metrics_on_evaluation_end` and the `on_evaluation_end` hook, would allow to inspect the epoch metrics in, for example, WandB, even if the trial gets terminated by Ray Tune. If one wants to create/modify the epoch metrics before they get logged, you still can use the `on_evaluation_epoch_end` hook.

### Alternatives

Introduce a new hook that is fired after the epoch metrics are logged. This hook could then be used by the tune reporter callback.

### Additional context

Not sure if there is an obvious reason for the current order, if so, please excuse my misunderstanding. 
If you think the change is desirable i would be more than happy to provide a pull request.

Keep up the great work, lightning is a really amazing library!

**Comment by justusschock** [comment]
Hi @dcfidalgo 

From what I saw, this should be possible to do.
@carmocca Is it possible to also log values in `on_evaluation_end`? I don't think so, right? So would you agree that these two things can be swapped?

cc @PyTorchLightning/core-contributors

**Comment by carmocca** [comment]
> @carmocca Is it possible to also log values in `on_evaluation_end`? I don't think so, right?

It is not:

https://github.com/PyTorchLightning/pytorch-lightning/blob/6be0a859db93989768f8bc845ee60e9c73f1e809/pytorch_lightning/trainer/connectors/logger_connector/callback_hook_validator.py#L223-L226

https://github.com/PyTorchLightning/pytorch-lightning/blob/6be0a859db93989768f8bc845ee60e9c73f1e809/pytorch_lightning/trainer/connectors/logger_connector/callback_hook_validator.py#L233-L236

> So would you agree that these two things can be swapped?

Yes
